### PR TITLE
[JENKINS-43889] Make sure Environments are always torn down (in AbstractBuild)

### PR DESCRIPTION
**Pull request by thomasgl-orange**
See [JENKINS-43889](https://issues.jenkins-ci.org/browse/JENKINS-43889).

TLDR for JENKINS-43889: Environments can be attached very early to an `AbstractBuild` (before SCM checkout), by either `BuildWrapper`s and `RunListener`s, but then they might never be torn down, if the build fails too early (in SCM checkout). A real-life example is the _SSH-Agent plugin_ leaving orphan `ssh-agent` processes running (when a Freestyle job, or Maven job, fails in SCM).

This PR contains three commits:
- first commit (5f324c7) adds a test case showing the issue in the case of a `SimpleBuildWrapper` set for pre-checkout execution (it shows a `Disposer` not being called in case of failure of SCM checkout)
- second commit (a5ab5dc) is the fix: it moves the teardown logic from the `AbtractBuildExecution#doRun` implementation (`BuildExecution#doRun` for Freestyle jobs) up to the `AbtractBuildExecution#run` method itself, in a `finally` block. To avoid tearing down environments twice in case a different `AbtractBuildExecution#doRun` implementation is at play (for instance in case of a Maven job), an extra `Environment` is added, which changes state when torn down, and serves as a marker.
- third commit (33ecdda) adds a test case to verify that the `Disposer` of a `SimpleBuildWrapper` is called _exactly_ once, including for a Maven build (to validate the above trick with the marker `Environment`).

**Update:** commits added after opening the PR:
- extraction of the tear down loop into its own method: 050f166
- utility method in `Result`, to get the worse of two nullable `Result`s: 2132275 + ee8002f **=> let me know if this doesn't deserve being in `Result`, and I should use a local private method instead**
- improve exceptions handling when tearing down environments: 625b783 + 0a44fe0

**Notes:**
- **this PR is easier to review when [ignoring whitespaces](https://github.com/jenkinsci/jenkins/pull/4517/files?w=1)** (because of the added `try/finally`)
- Before this change, the `AbstractBuild#buildEnvironments` list was only initialized in `createLauncher()` after the wrappers' `decorateLauncher()` had been called. Which means these methods were not able to contribute environments (I think). It is now done a bit earlier, meaning they can. I don't know if it's an issue.
- The code for calling `tearDown` on environments in `finally` block is a bit different from what was previously in `BuildExecution#doRun`: is catches exceptions, and tries hard to complete iteration on all environments. It is closer to what's done [in the Maven plugin](https://github.com/jenkinsci/maven-plugin/blob/master/src/main/java/hudson/maven/MavenBuild.java#L935). And in case of an `InterruptedException`, I was not sure what to do TBH: ~~I went for finishing the iteration and then interrupting the thread again (and setting result to FAILED, like for any other exception)~~ after some reviewers feedback, I went for re-throwing the (first) caught `InterruptedException`, and let the caller (`Run#execute(RunExecution)`) deal with it (which will result in an `ABORTED` build result).

### Proposed changelog entries

* JENKINS-43889, bug: fix environments not always being torn down ([issue 43889](https://issues.jenkins-ci.org/browse/JENKINS-43889))
* JENKINS-43889, rfe: Developer: new static utility method `Result#combine(Result,Result)` to get the worst of two (nullable) build results

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [X] JIRA issue is well described
- [X] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [X] Appropriate autotests or explanation to why this change has no tests
- [ ] ~~For dependency updates: links to external changelogs and, if possible, full diffs~~

### Desired reviewers

Maybe @jglick? (as author of `SimpleBuildWrapper(Test)`)

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [x] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [x] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [x] Proper changelog labels are set so that the changelog can be generated automatically
- [x] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

### CODE REVIEW

**Comment by thomasgl-orange** [comment]
:man_facepalming: 
Thanks @daniel-beck. Now fixed in code and PR too.

File: core/src/main/java/hudson/model/AbstractBuild.java [hunk 1]
```diff
@@ -518,6 +559,20 @@ public Result run(@Nonnull BuildListener listener) throws Exception {
             return result;
         }
 
+        /**
+         * An {@link Environment} which does nothing, but change state when it gets torn down. Used in
+         * {@link AbstractBuildExecution#run(BuildListener)} to detect whether environments have yet to be torn down,
+         * or if it has been done already (in the {@link AbstractBuildExecution#doRun(BuildListener)} implementation).
+         */
+        private class TearDownCheckEnvironment extends Environment {
```
Comment (res0nance) [inline core/src/main/java/hudson/model/AbstractBuild.java]:
```suggestion
        private static class TearDownCheckEnvironment extends Environment {
```

File: core/src/main/java/hudson/model/AbstractBuild.java [hunk 2]
```diff
@@ -459,49 +459,90 @@ public Result run(@Nonnull BuildListener listener) throws Exception {
...
+                        final Environment environment = buildEnvironments.get(i);
+                        try {
+                            if (!environment.tearDown(AbstractBuild.this, listener)) {
+                                tearDownFailed = true;
+                            }
+                        } catch (IOException | InterruptedException | RuntimeException e) {
+                            tearDownFailed = true;
+                            // exceptions are ignored to give a chance to all environments to tear down
+                            listener.error("Unable to tear down: " + e.getMessage());
+                            e.printStackTrace(listener.getLogger());
```
Comment (res0nance) [inline core/src/main/java/hudson/model/AbstractBuild.java]:
```suggestion
                            Functions.printStackTrace(e,listener.getLogger());
```
Core has something close to this i forgot what it was exactly

File: core/src/main/java/hudson/model/AbstractBuild.java [hunk 3]
```diff
@@ -459,49 +459,90 @@ public Result run(@Nonnull BuildListener listener) throws Exception {
...
+                            if (e instanceof InterruptedException) {
+                                // don't forget we've been interrupted
+                                tearDownInterrupted = true;
```
Comment (res0nance) [inline core/src/main/java/hudson/model/AbstractBuild.java]: I'm not sure if this is the right approach since being interrupted usually means to exit ASAP

Comment (thomasgl-orange) [reply core/src/main/java/hudson/model/AbstractBuild.java]: I don't have a set opinion on what to do after the tear-down loop in case we've caught an `InterruptedException`. I've chosen to `Thread.currentThread().interrupt()`, but maybe we could as well raise a new `InterruptedException`, or maybe simply set an appropriate build status.

But I really think that completing the iteration on remaining environments, even in case of an `InterruptedException`, is the right thing to do. Leaving temporary files behind, or orphan processes, by not tearing down the remaining environments seems "wrong" to me (when it can be avoided). I think of multiple build wrappers (environments) not as a list (although that's how it is implemented), but rather as nested try/finally blocks.

Let's wait for more reviews before changing this code...

File: core/src/main/java/hudson/model/AbstractBuild.java (hunk 1 above)
Comment (thomasgl-orange) [reply core/src/main/java/hudson/model/AbstractBuild.java]: Actually, now I remember... It's not static because it is nested in a non-static nested class (`AbstractBuild.AbstractBuildExecution`). Do you think I should move it?

File: core/src/main/java/hudson/model/AbstractBuild.java [hunk 4]
```diff
@@ -459,49 +459,90 @@ public Result run(@Nonnull BuildListener listener) throws Exception {
...
+                }
 
-            getProject().getScmCheckoutStrategy().preCheckout(AbstractBuild.this, launcher, this.listener);
-            getProject().getScmCheckoutStrategy().checkout(this);
+                getProject().getScmCheckoutStrategy().preCheckout(AbstractBuild.this, launcher, this.listener);
+                getProject().getScmCheckoutStrategy().checkout(this);
 
-            if (!preBuild(listener,project.getProperties()))
-                return Result.FAILURE;
+                if (!preBuild(listener,project.getProperties()))
```
Comment (varyvol) [inline core/src/main/java/hudson/model/AbstractBuild.java]: Maybe a good opportunity to include curly brackets here.

File: core/src/main/java/hudson/model/AbstractBuild.java [hunk 5]
```diff
@@ -459,49 +459,90 @@ public Result run(@Nonnull BuildListener listener) throws Exception {
...
 
-            Result result = doRun(listener);
+                result = doRun(listener);
+            } finally {
+                if (!tearDownMarker.tornDown) {
+                    // looks like environments are not torn down yet, do it now (in reverse order)
+                    boolean tearDownFailed = false;
+                    boolean tearDownInterrupted = false;
+                    for (int i = buildEnvironments.size() - 1; i >= 0; i--) {
+                        final Environment environment = buildEnvironments.get(i);
```
Comment (varyvol) [inline core/src/main/java/hudson/model/AbstractBuild.java]: for (Environment  environment : buildEnvironments) {

File: core/src/main/java/hudson/model/AbstractBuild.java [hunk 6]
```diff
@@ -459,49 +459,90 @@ public Result run(@Nonnull BuildListener listener) throws Exception {
...
 
-            if (!preBuild(listener,project.getProperties()))
-                return Result.FAILURE;
+                if (!preBuild(listener,project.getProperties()))
+                    return Result.FAILURE;
 
-            Result result = doRun(listener);
+                result = doRun(listener);
+            } finally {
+                if (!tearDownMarker.tornDown) {
```
Comment (varyvol) [inline core/src/main/java/hudson/model/AbstractBuild.java]: Just a suggestion but maybe this would be nicer to read/follow if moved to an independent method, the current method is already long enough.

File: core/src/main/java/hudson/model/AbstractBuild.java (hunk 4 above)
Comment (thomasgl-orange) [reply core/src/main/java/hudson/model/AbstractBuild.java]: If you ignore whitespaces, I'm not really modifying these lines of code:
https://github.com/jenkinsci/jenkins/pull/4517/files/15193af21d387218ed248182b2ef3e398ad0a593?utf8=%E2%9C%93&diff=split&w=1
So I'd rather not change the code style, I think it would be unrelated to the actual modifications from this PR.

File: core/src/main/java/hudson/model/AbstractBuild.java (hunk 5 above)
Comment (thomasgl-orange) [reply core/src/main/java/hudson/model/AbstractBuild.java]: I'm iterating in reverse order (not sure if there are actual cases where it matters, but that's what was done previously).

Comment (MRamonLeon) [reply core/src/main/java/hudson/model/AbstractBuild.java]: I think he didn't do so to do it in reverse order

File: core/src/main/java/hudson/model/AbstractBuild.java [hunk 7]
```diff
@@ -459,49 +459,90 @@ public Result run(@Nonnull BuildListener listener) throws Exception {
...
+                            }
+                        }
+                    }
+                    // report any error while tearing down an Environment as a build failure
+                    if (tearDownFailed) {
+                        result = Result.FAILURE;
+                    }
+                    if (tearDownInterrupted) {
+                        // don't forget we've been interrupted
+                        Thread.currentThread().interrupt();
```
Comment (MRamonLeon) [inline core/src/main/java/hudson/model/AbstractBuild.java]: Isn't it better to throw an interrupted exception here?

File: core/src/main/java/hudson/model/AbstractBuild.java (hunk 6 above)
Comment (thomasgl-orange) [reply core/src/main/java/hudson/model/AbstractBuild.java]: I agree, thanks for the suggestion. 050f1661db

File: core/src/main/java/hudson/model/AbstractBuild.java (hunk 4 above)
Comment (varyvol) [reply core/src/main/java/hudson/model/AbstractBuild.java]: Fair enough!

File: core/src/main/java/hudson/model/AbstractBuild.java (hunk 5 above)
Comment (varyvol) [reply core/src/main/java/hudson/model/AbstractBuild.java]: Oh right, sorry I didn't notice!

**Review by varyvol — APPROVED** [review]

File: core/src/main/java/hudson/model/AbstractBuild.java [hunk 8]
```diff
@@ -459,49 +459,90 @@ public Result run(@Nonnull BuildListener listener) throws Exception {
...
+                            Functions.printStackTrace(e, listener.getLogger());
+                            if (e instanceof InterruptedException) {
+                                // don't forget we've been interrupted
+                                tearDownInterrupted = true;
+                            }
+                        }
+                    }
+                    // report any error while tearing down an Environment as a build failure
+                    if (tearDownFailed) {
+                        result = Result.FAILURE;
```
[the remark this answers is not in the document]
Comment (thomasgl-orange) [reply core/src/main/java/hudson/model/AbstractBuild.java]: Regarding `UNSTABLE` on `tearDown` exception, I tend to agree, but it was `FAILURE` prior to this PR (in the `BuildExecution#doRun` implementation), so I don't really want to change this behavior here. It goes beyond the fix for "environments not always being torn down", and I don't know if some environments implementations rely on this behavior (throw exception, instead of returning false, to signal a fatal build error).

Regarding `InterruptedException`, after reading how it's handled in `Run#execute(RunExecution)` (and `Executor#abortResult()`), I've finally decided to re-throw an `InterruptedException` (the first one which has been caught) rather than calling `Thread#interrupt()`, and let it be handled by the caller (which will set an `ABORTED` result in the end, but with a few subtleties regarding the cause of interruption).

That's 625b783b9f.

File: core/src/main/java/hudson/model/AbstractBuild.java (hunk 7 above)
Comment (thomasgl-orange) [reply core/src/main/java/hudson/model/AbstractBuild.java]: Yes, see my answer to your previous comment, and 625b783b9f.

File: core/src/main/java/hudson/model/AbstractBuild.java (hunk 3 above)
Comment (thomasgl-orange) [reply core/src/main/java/hudson/model/AbstractBuild.java]: > I don't have a set opinion on what to do after the tear-down loop in case we've caught an InterruptedException.

I've now changed my code to re-throw an `InterruptionException`. See https://github.com/jenkinsci/jenkins/pull/4517#discussion_r385416603 and 625b783.

File: core/src/main/java/hudson/model/AbstractBuild.java [hunk 9]
```diff
@@ -459,49 +460,64 @@ public Result run(@Nonnull BuildListener listener) throws Exception {
...
-                    if (!assignedLabels.isEmpty()) {
-                        boolean first = true;
-                        for (LabelAtom label : assignedLabels) {
-                            if (first) {
-                                listener.getLogger().print(" (");
-                                first = false;
-                            } else {
-                                listener.getLogger().print(' ');
+            Result result = null;
+            buildEnvironments = new ArrayList<>();
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/AbstractBuild.java]: This historical implementation just asks for concurrency issues at some point. Not blocking this change, because it does not make it any worse

File: core/src/main/java/hudson/model/AbstractBuild.java [hunk 10]
```diff
@@ -459,49 +460,64 @@ public Result run(@Nonnull BuildListener listener) throws Exception {
...
-                return Result.FAILURE;
+                if (!preBuild(listener,project.getProperties()))
+                    return Result.FAILURE;
 
-            Result result = doRun(listener);
+                result = doRun(listener);
+            } finally {
+                if (!tearDownMarker.tornDown) {
+                    // looks like environments are not torn down yet, do it now (might affect the build result)
+                    result = Result.combine(result, tearDownBuildEnvironments(listener));
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/model/AbstractBuild.java]: InterruptedException in `tearDownBuildEnvironments()` will bypass this logic and cause loss of the original build result information. I doubt this is what we want

File: core/src/main/java/hudson/model/AbstractBuild.java (hunk 1 above)
Comment (oleg-nenashev) [reply core/src/main/java/hudson/model/AbstractBuild.java]: Would be better to avoid non-static nested classes when possible. Better safe than sorry

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
I do not see any implementation issues except the loss of information when `InterruptedException `is thrown by the teardown logic. It looks like the results should be combined taking handling of other negative scenarios. The rest of the code looks good to me.

File: core/src/main/java/hudson/model/AbstractBuild.java (hunk 1 above)
Comment (thomasgl-orange) [reply core/src/main/java/hudson/model/AbstractBuild.java]: Okay,that's 462104a

File: core/src/main/java/hudson/model/AbstractBuild.java (hunk 10 above)
Comment (thomasgl-orange) [reply core/src/main/java/hudson/model/AbstractBuild.java]: That's intentional, see https://github.com/jenkinsci/jenkins/pull/4517#discussion_r385416603
My understanding of how it works today is that an `InterruptedException` during an environment `tearDown()` would go up to `Run#execute(RunExecution)`, where it is handled in a "special" way (`Executor#abortResult()`) to take user interruption cause into account. In the end, the result is `ABORTED`, effectively bypassing the result of the build steps.
I don't really want to change/rewrite that, it's not what this pull request is about. I've tried to change things "just enough" to make sure all environments are torn down even in this case (I'll still try to tear down remaining environments before re-throwing the exception), but no more than that.

I've added a test case (83c2602) which shows that:
1. the result is `ABORTED` in this case,
2. remaining environments are torn down.

Executing the same test without the rest of this pull-request still passes 1 (the `ABORTED` assertion), and fails on 2. (see thomasgl-orange/jenkins@dc71032fd7 for a cherry pick of 83c2602 on top of master).

**Review by MRamonLeon — APPROVED** [review]

**Comment by varyvol** [comment]
@thomasgl-orange would you mind resolving the merge conflicts so this can move forward?

**Comment by thomasgl-orange** [comment]
@varyvol thanks for the heads-up, I've solved the merge conflicts (s/Nonnull/NonNull/).

What about the `squash-merge-me` label? Does it mean I should push-force cleaned up commits here? Or is it up to the core maintainers to squash things when merging the PR?

**Comment by varyvol** [comment]
@thomasgl-orange no, better to avoid push-force. It's something for maintainers to take care of.

File: core/src/main/java/hudson/model/AbstractBuild.java (hunk 9 above)
Comment (thomasgl-orange) [reply core/src/main/java/hudson/model/AbstractBuild.java:464]: > I do not see any implementation issues except the loss of information when `InterruptedException `is thrown by the teardown logic. It looks like the results should be combined taking handling of other negative scenarios. The rest of the code looks good to me.

@oleg-nenashev: Are you ok with https://github.com/jenkinsci/jenkins/pull/4517#discussion_r389186582 or are you still requesting changes in the handling of `InterruptedException` from teardown?

One minor, but simple, improvement I can think of would be https://github.com/thomasgl-orange/jenkins/commit/2e9b34260a4000d12f0b769324c2d6499d457aec to log the `Result` of the `doRun` phase, which is the information we're about to lose when re-throwing the `InterruptedException`. Would that help, or is there something else you had in mind?

**Comment by alecharp** [comment]
@oleg-nenashev could you have a look at this one, as @thomasgl-orange did answer you. Thanks!

**Comment by MRamonLeon** [comment]
@oleg-nenashev I've set the _ready-for-merge_ label here as it seems you don't have any blocker for this to get merged. If there is still pending changes from your side, please let us know. Thanks
